File: services/renderer-go/renderer/renderer.go
package renderer

import (
	"bytes"
	"context"
	"html/template"
	"regexp"
)

var standAloneUrlRE = regexp.MustCompile(`[^(\(.*\)\[)]https?://[^\s]+[^\]]`)
var exactUrlRE = regexp.MustCompile(`https?://[^\s]+`)
var linkTmpl = template.Must(template.New("link").Parse(`<a href="{{.}}">{{.}}</a>`))

// Render は受け取った文書を HTML に変換する
func Render(ctx context.Context, src string) (string, error) {
	// TODO: これはサンプル実装 (URL の自動リンク) です
	html := standAloneUrlRE.ReplaceAllStringFunc(src, func(standAloneUrl string) string {
		return exactUrlRE.ReplaceAllStringFunc(standAloneUrl, func(exactUrl string) string {
			var w bytes.Buffer
			err := linkTmpl.ExecuteTemplate(&w, "link", exactUrl)
			if err != nil {
				return exactUrl
			}
			return w.String()
		})
	})
	return html, nil
}

File: services/renderer-ts/src/renderer.ts
/**
 * 受け取った文書を HTML に変換する
 */
export async function render(src: string): Promise<string> {
  // TODO: これはサンプル実装 (URL の自動リンク) です
  const html = src.replace(/https?:\/\/[^\s]+/g, (url) => `<a href="${encodeURI(url)}">${url}</a>`);
  return html;
}
